windows/src/engine/keyman32/kmprocessactions.cpp

```diff
@@ -0,0 +1,135 @@
+/*
+ * Keyman is copyright (C) SIL International. MIT License.
+ *
+ * Description: Used for apply core processor actions to
+ *              Keyman for Windows engine.
+ */
+#include "pch.h"
+
+static BOOL processUnicodeChar(AITIP* app, const km_kbp_action_item* actionItem) {
+  if (Uni_IsSMP(actionItem->character)) {
+    app->QueueAction(QIT_CHAR, (Uni_UTF32ToSurrogate1(actionItem->character)));
+    app->QueueAction(QIT_CHAR, (Uni_UTF32ToSurrogate2(actionItem->character)));
+  }
+  else {
+    app->QueueAction(QIT_CHAR, actionItem->character);
+  }
+  return TRUE;
+}
+
+static BOOL processMarker(AITIP* app, const km_kbp_action_item* actionItem) {
+  app->QueueAction(QIT_DEADKEY, (DWORD)actionItem->marker);
+  return TRUE;
+}
+
+static BOOL processAlert(AITIP* app) {
+  app->QueueAction(QIT_BELL, 0);
+  return TRUE;
+}
+
+static BOOL processBack(AITIP* app, const km_kbp_action_item* actionItem) {
+  if (actionItem->backspace.expected_type == KM_KBP_BT_MARKER) {
+    app->QueueAction(QIT_BACK, BK_DEADKEY);
+  } else /* actionItem->backspace.expected_type == KM_KBP_BT_CHAR, KM_KBP_BT_UNKNOWN */ {
+    app->QueueAction(QIT_BACK, 0);
+  }
+  return TRUE;
+}
+
+static BOOL processPersistOpt(
+  const km_kbp_action_item* actionItem,
+  km_kbp_state* keyboardState,
+  LPINTKEYBOARDINFO activeKeyboard
+) {
+  if (actionItem->option != NULL)
+  {
+    // Allocate for 1 option plus 1 pad struct of 0's for KM_KBP_IT_END
+    km_kbp_option_item keyboardOpts[2] = { 0 };
+
+    km_kbp_cp* cp = new km_kbp_cp[sizeof(actionItem->option->key) + 1];
+    wcscpy_s(reinterpret_cast<LPWSTR>(cp), sizeof(actionItem->option->key) + 1, reinterpret_cast<LPCWSTR>(actionItem->option->key));
+    keyboardOpts[0].key = cp;
+
+    cp = new km_kbp_cp[sizeof(actionItem->option->value) + 1];
+    wcscpy_s(reinterpret_cast<LPWSTR>(cp), sizeof(actionItem->option->value) + 1, reinterpret_cast<LPCWSTR>(actionItem->option->value));
+    keyboardOpts[0].value = cp;
```

Comment:
Do we need to make a copy of these strings, or can we just do `keyboardOptions[0].key = actionItem->option->key;` etc?

Looking at the implementation of `km_kbp_state_options_update`, it only reads the strings. So we shouldn't need to be making the copy.